src/daemon/backends/connman_service.cpp

```diff
@@ -0,0 +1,298 @@
+        constexpr char STATE_STR_CONFIGURATION[] = "configuration";
+        constexpr char STATE_STR_READY[] = "ready";
+        constexpr char STATE_STR_DISCONNECT[] = "disconnect";
+        constexpr char STATE_STR_ONLINE[] = "online";
+
+        ConnManService::Type type_from_string(const Glib::ustring &str)
+        {
+            if (str == TYPE_STR_BLUETOOTH)
+                return ConnManService::Type::BLUETOOTH;
+
+            if (str == TYPE_STR_ETHERNET)
+                return ConnManService::Type::ETHERNET;
+
+            if (str == TYPE_STR_WIFI)
+                return ConnManService::Type::WIFI;
+
+            return ConnManService::Type::UNKNOWN;
+        }
+
+        Glib::ustring type_to_string(ConnManService::Type type)
+        {
+            switch (type) {
+            case ConnManService::Type::UNKNOWN:
+                break;
+            case ConnManService::Type::BLUETOOTH:
+                return TYPE_STR_BLUETOOTH;
+            case ConnManService::Type::ETHERNET:
+                return TYPE_STR_ETHERNET;
+            case ConnManService::Type::WIFI:
+                return TYPE_STR_WIFI;
+            }
+            return "unknown";
+        }
+
+        ConnManService::State state_from_string(const Glib::ustring &str)
+        {
+            if (str == STATE_STR_IDLE)
+                return ConnManService::State::IDLE;
+
+            if (str == STATE_STR_FAILURE)
+                return ConnManService::State::FAILURE;
+
+            if (str == STATE_STR_ASSOCIATION)
+                return ConnManService::State::ASSOCIATION;
+
+            if (str == STATE_STR_CONFIGURATION)
+                return ConnManService::State::CONFIGURATION;
+
+            if (str == STATE_STR_READY)
+                return ConnManService::State::READY;
+
+            if (str == STATE_STR_DISCONNECT)
+                return ConnManService::State::DISCONNECT;
+
+            if (str == STATE_STR_ONLINE)
+                return ConnManService::State::ONLINE;
+
+            g_warning(R"(Received unknown ConnMan service state "%s", defaulting to "idle")",
+                      str.c_str());
+
+            return ConnManService::State::IDLE;
+        }
+
+        std::optional<ConnManService::State> state_from_string(
+            const std::optional<Glib::ustring> &str)
+        {
+            if (!str)
+                return {};
+            return state_from_string(*str);
+        }
+
+        ConnManService::Strength strength_from_uint8(std::uint8_t i)
+        {
+            constexpr std::uint8_t MAX = 100;
+            return i <= MAX ? i : MAX;
+        }
+
+        std::optional<ConnManService::Strength> strength_from_uint8(std::optional<std::uint8_t> i)
+        {
+            if (!i)
+                return {};
+            return strength_from_uint8(*i);
+        }
+
+        template <typename T>
+        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
```

Comment:
This function is defined in 3 different files within anonymous namespace. Can this be made as a common function as it is being used in many files?

Reply:
The log messages are unique in each one and it is only a thin wrapper around Glib::VariantBase::dynamic_cast() since it throws. But perhaps it makes sense to add the wrapping in a helper function in src/common/. Would still keep these local functions though since I think it makes sense to have different log messages (would just be changed to check return value instead of try/catch and g_warning would be kept here in local helper function).

Reply:
Ideally there should be a function for this in glibmm... A version of Glib::Variant::cast_dynamic() that does not throw and returns the value directly instead of the value wrapped in a variant.

Reply:
Tested adding a helper function here: https://github.com/Pelagicore/connectivity-manager/pull/1/commits/d4c4a34a171f78972c0fe08bb09d5105c2d9af5d

I do not think it is worth it. Still want the specific warning messages. So this:

```C++
        template <typename T>
        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
                                            const Glib::ustring &name)
        {
            try {
                return Glib::VariantBase::cast_dynamic<Glib::Variant<T>>(variant).get();
            } catch (const std::bad_cast &) {
                g_warning("Invalid type %s for ConnMan service property \"%s\"",
                          variant.get_type_string().c_str(),
                          name.c_str());
            }
            return {};
        }
```

Becomes this:

```C++
        template <typename T>
        std::optional<T> value_from_variant(const Glib::VariantBase &variant,
                                            const Glib::ustring &name)
        {
            std::optional<T> ret = Common::variant_get<T>(variant);

            if (!ret)
                g_warning("Invalid type %s for ConnMan service property \"%s\"",
                          variant.get_type_string().c_str(),
                          name.c_str());

            return ret;
        }
```

I do not like e.g. adding a "warning string and printf arguments" to Common::variant_get()... too specific for a generic function and doing the "printf argument"-dance is not fun. Sometimes duplicating code (tm) is just the best thing to do. I think this is one of those times.

Or do you perhaps have any other suggestion for how to improve it?

Reply:
Just wondering why not just return the value and print warning where this function is used based on the return value?

Reply:
But I guess you don't want to print warning on every usage of this function. That would also be a type of duplication. That's why you wrapped variant_get inside value_from_variant.

Reply:
If it is possible may be the ideal would be to modify the generator to generate code that returns typed values (As you have suggested a couple of times before). But that would be a different user story.

Reply:
> But I guess you don't want to print warning on every usage of this function. That would also be a type of duplication. 

Exactly.

> But that would be a different user story.

A different story and a large one at that, I think. Since "structs" does not exist in D-Bus one would have to add custom annotations. I opened https://gitlab.freedesktop.org/dbus/dbus/issues/141 in D-Bus a couple of years ago and linked to an example (link does not work anymore though... it looks like).

Reply:
But also, in this case the type in the D-Bus specification is a variant. So perhaps the "structs with names on members" part is not relevant. :)

Reply:
Reverted https://github.com/Pelagicore/connectivity-manager/pull/1/commits/d4c4a34a171f78972c0fe08bb09d5105c2d9af5d and will remove it completely when squashing.